strider/trapi.py

```diff
@@ -247,10 +247,58 @@ def filter_message(
                                         ] = message.knowledge_graph.edges[edge_id]
 
     message.results = kept_results
-    message.knowledge_graph.edges = kept_knowledge_graph.edges
+    if message.knowledge_graph:
+        message.knowledge_graph.edges = kept_knowledge_graph.edges
+    else:
+        message.knowledge_graph = kept_knowledge_graph
     message.auxiliary_graphs = kept_aux_graphs
 
 
+def clean_query_id(
+    message: Message,
+    curie_map: dict,
+    kp_id: str,
+    logger: logging.Logger = logging.getLogger(),
+):
+    """Remove any errant query_ids from result node_bindings."""
+    for result in message.results or []:
+        for qnode_id, node_bindings in result.node_bindings.items():
+            for node_binding in node_bindings:
+                if (
+                    node_binding.query_id is not None
+                    and node_binding.query_id
+                    not in message.query_graph.nodes[qnode_id].ids
+                ):
+                    # if query id exists and isn't in qgraph
+                    if node_binding.id in message.query_graph.nodes[qnode_id].ids:
+                        # if kgraph id in qgraph, then remove query id
+                        # this was probably a preferred prefix
+                        logger.debug(
+                            f"Removing query_id {node_binding.query_id} because kgraph id is in qgraph."
+                        )
+                        node_binding.query_id = None
+                    else:
+                        # both kgraph id and query id not in qgraph,
+                        # need to figure out where query id came from
+                        if node_binding.query_id in curie_map:
+                            # query id is an equivalent identifier that we know about
+                            for eq_id in curie_map[node_binding.query_id].identifiers:
+                                if eq_id in message.query_graph.nodes[qnode_id].ids:
+                                    # found a match
+                                    logger.info(
+                                        f"Changing query_id from {node_binding.query_id} to {eq_id}."
+                                    )
+                                    node_binding.query_id = eq_id
```

Comment:
If merging is the reason why this is necessary, maybe instead of changing the query_id to the node_id when they are equivalent, we should get rid of the query_id entirely?

Reply:
I don't actually check that the node_binding `id` and `query_id` are equivalent, only if either is in the qgraph. Where you put this comment, I'm checking equivalent identifiers of the query_id to see if anything matches in the qgraph. I don't think we can just get rid of the query_id, because we do need it if the kgraph node is a subclass or something.